Improves UART reading performance

SuGlider opened the pull request:
## Description of Change
Improves UART reading more than 1 bytes, using `HardwareSerial::read(uint8_t *buff, size_t len)`

## Tests scenarios
Tested with ESP32, ESP32S2, ESP32S3, ESP32C3 using this sketch and a Python script runing on a computer.
See #7505 

General Testing Sketch:
``` cpp
//Testing by sending "0123456789abcdefghijklmnopqrstuvwxyz" using Serial Monitor

void setup() {
  Serial.begin(115200);
  Serial.println("\nSend anything to Serial... testing read() and peek()\n");
}

void loop() {
  if (Serial.available()) {
    Serial.printf("Serial has %d bytes available.\n", Serial.available());
    Serial.printf("Serial PEEK is %c\n", Serial.peek());
    int i = Serial.read();
    Serial.printf("Serial Read 1 byte after Peek: %d ['%c']\n", i, i);
    i = Serial.read();
    Serial.printf("Serial Read again 1 byte after First Read: %d ['%c']\n", i, i);
    Serial.println("Serial Read in blocks of 10 bytes:");
    i = Serial.peek();
    Serial.printf("Serial PEEK before trying to read a chunk of data is %d ['%c']\n", i, i);
    uint8_t buf[11];
    size_t len = 0;
    while (len = Serial.read(buf, 10)) {
    //while (Serial.available()) {
      buf[len] = '\0';
      Serial.println((char*)buf);
      i = Serial.peek();
      Serial.printf("Serial PEEK is %d ['%c']\n", i, i);
    }
    Serial.println("===============================================");
  }
}
```

Expected OUPUT when sending 0123456789abcdefghijklmnopqrstuvwxyz
```
Serial has 36 bytes available.
Serial PEEK is 0
Serial Read 1 byte after Peek: 48 ['0']
Serial Read again 1 byte after First Read: 49 ['1']
Serial Read in blocks of 10 bytes:
Serial PEEK before trying to read a chunk of data is 50 ['2']
23456789ab
Serial PEEK is 99 ['c']
cdefghijkl
Serial PEEK is 109 ['m']
mnopqrstuv
Serial PEEK is 119 ['w']
wxyz
Serial PEEK is -1 ['⸮']
===============================================
```

## Related links
Fixes #7505
Fixes #7474  

May Fix #6644
May Fix #6727

TD-er commented on cores/esp32/esp32-hal-uart.c:
```diff
@@ -326,6 +326,34 @@ uint32_t uartAvailableForWrite(uart_t* uart)
+    }
+
+    if (size > 0) {
+       int len = uart_read_bytes(uart->num, buffer, size, pdMS_TO_TICKS(timeout_ms));
```
Since you're calling it with 0 msec timeout, I wonder what does 0 timeout mean here?
Is it "no wait" or "wait forever" ?
It isn't defined here: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/uart.html

SuGlider replied:
Zero means, no waiting. If the bytes are not available, it will not wait and return immediately.
portMAX_DELAY is used as the "waiting for ever".

TD-er replied:
thanks for the update!

mrengineer7777 reviewed (changes requested):
In esp32-hal-uart.c->uartRead(), `int len = uart_read_bytes(uart->num, &c, 1, 20 / portTICK_RATE_MS);` can return negative.

mrengineer7777 reviewed (commented):
PR drops call to available().  Likely OK, but need to test both `HardwareSerial::read(void)` and `HardwareSerial::read(uint8_t *buffer, size_t size)` when no data available to see if there are any side effects with calling IDF function uart_read_bytes() with 0 timeout.

PR looks promising.  I see no other issues.

SuGlider commented:
> PR drops call to available(). Likely OK, but need to test both `HardwareSerial::read(void)` and `HardwareSerial::read(uint8_t *buffer, size_t size)` when no data available to see if there are any side effects with calling IDF function uart_read_bytes() with 0 timeout.
> 
> PR looks promising. I see no other issues.

I tested it using the example sketch of this PR.
I sent just "0", "01", "012" and "0123". 
All of them try to read or peek from empty RX buffer at different moments. 
Everything works as expected.

P-R-O-C-H-Y commented:
@SuGlider I am getting this output from the sketch on ESP32-S2 (did not test other chips yet).

```
Serial has 36 bytes available.
Serial PEEK is 0
Serial Read 1 byte after Peek: 48 ['0']
Serial Read again 1 byte after First Read: 49 ['1']
Serial Read in blocks of 10 bytes:
Serial PEEK before trying to read a chunk of data is 50 ['2']
Serial PEEK is 99 ['c']
23456789ab
Serial PEEK is 109 ['m']
cdefghijkl
Serial PEEK is 119 ['w']
mnopqrstuv
Serial PEEK is -1 ['⸮']
wxyz
===============================================
```

Is that right? If yes maybe update the expected OUTPUT :)

P-R-O-C-H-Y commented:
Also did a quick test of the performance of the reading as the sketch is written (only Serial.prints removed).

Before changes reading 36 bytes took 316-325 us.
After changes reading 36 bytes took 118-126 us.

Seems a really nice speed improvement. Great PR @SuGlider

SuGlider commented:
New commit e0723028494de5ad6349a039bdd5a190fc5c61ab overrides `Stream::readBytes()`  to make it faster with IDF.

SuGlider commented:
> @SuGlider I am getting this output from the sketch on ESP32-S2 (did not test other chips yet).
> 
> ```
> Serial has 36 bytes available.
> Serial PEEK is 0
> Serial Read 1 byte after Peek: 48 ['0']
> Serial Read again 1 byte after First Read: 49 ['1']
> Serial Read in blocks of 10 bytes:
> Serial PEEK before trying to read a chunk of data is 50 ['2']
> Serial PEEK is 99 ['c']
> 23456789ab
> Serial PEEK is 109 ['m']
> cdefghijkl
> Serial PEEK is 119 ['w']
> mnopqrstuv
> Serial PEEK is -1 ['⸮']
> wxyz
> ===============================================
> ```
> 
> Is that right? If yes maybe update the expected OUTPUT :)

Thanks. Example and Output are now updated.

mrengineer7777 commented:
@SuGlider See my review comments above, especially on readBytes().

SuGlider commented:
> @SuGlider See my review comments above, especially on readBytes().

This one https://github.com/espressif/arduino-esp32/pull/7525#pullrequestreview-1195948195?

SuGlider commented:
> In esp32-hal-uart.c->uartRead(), `int len = uart_read_bytes(uart->num, &c, 1, 20 / portTICK_RATE_MS);` can return negative.

The code already treats negative return:

```
if (size > 0) {
       int len = uart_read_bytes(uart->num, buffer, size, pdMS_TO_TICKS(timeout_ms));
       if (len < 0) len = 0;  // error reading UART   <<<=======================Here
       bytes_read += len;
    }
```

mrengineer7777 commented:
@SuGlider Apologies, Github is being weird about my review notes and it showing review as pending (but not posted?).  I'll cancel the review, look over the current code and just comment here.

mrengineer7777 commented:
On cores/esp32/HardwareSerial.h 'readBytes(char *buffer, size_t length)`, should this be uint8_t instead of char? recommend: `return readBytes((uint8_t *)buffer, length);`

On cores/esp32/esp32-hal-uart.c `uint8_t uartRead(uart_t* uart)` (not currently modified by PR) uart_read_bytes() can return negative.  Recommend changing `if (len == 0) {` to `if(len <= 0) {`.  uartPeek() may have the same problem.

On cores/esp32/esp32-hal-uart.c `uint8_t uartRead(uart_t* uart)` I recommend the following comments on the function:
```
//Reads one byte from uart.
//Check uartAvailable() > 0 before calling uartRead().
```
Also consider marking uartRead() as `\\Deprecated. Use uartReadBytes() instead.`

SuGlider commented:
> On cores/esp32/HardwareSerial.h `readBytes(char *buffer, size_t length)`, should this be uint8_t instead of char? recommend: `return readBytes((uint8_t *)buffer, length);`
> 
> On cores/esp32/esp32-hal-uart.c `uint8_t uartRead(uart_t* uart)` (not currently modified by PR) uart_read_bytes() can return negative. Recommend changing `if (len == 0) {` to `if(len <= 0) {`. uartPeek() may have the same problem.
> 
> On cores/esp32/esp32-hal-uart.c `uint8_t uartRead(uart_t* uart)` I recommend the following comments on the function:
> 
> ```
> //Reads one byte from uart.
> //Check uartAvailable() > 0 before calling uartRead().
> ```
> 
> Also consider marking uartRead() as `//Deprecated. Use uartReadBytes() instead.`

It is a bit confusing to review code this way. Do you mind using the traditional way as below:

When reviewing it within GitHub, you can use 2 buttons:

![image](https://user-images.githubusercontent.com/65977697/206004785-cbf0665f-a0cc-4708-b312-de80ddcba47f.png)

The Red Arrow will insert the review immediately.
The Yellow Arrow, adds the review, but it will be published only when all the review is submitted, by pressing the Yellow marked button below:

![image](https://user-images.githubusercontent.com/65977697/206005222-0cb70cb4-fc46-4f4d-9dbc-ea815136a982.png)

SuGlider commented:
> On cores/esp32/HardwareSerial.h `readBytes(char *buffer, size_t length)`, should this be uint8_t instead of char? recommend: `return readBytes((uint8_t *)buffer, length);`

Yes, correct. Fixed now.

> On cores/esp32/esp32-hal-uart.c `uint8_t uartRead(uart_t* uart)` (not currently modified by PR) uart_read_bytes() can return negative. Recommend changing `if (len == 0) {` to `if(len <= 0) {`. uartPeek() may have the same problem.
> 

Yes, noted. Both functions are now fixed.

> On cores/esp32/esp32-hal-uart.c `uint8_t uartRead(uart_t* uart)` I recommend the following comments on the function:
> 
> ```
> //Reads one byte from uart.
> //Check uartAvailable() > 0 before calling uartRead().
> ```
> 
> Also consider marking uartRead() as `//Deprecated. Use uartReadBytes() instead.`

Actually the functions exposed in esp32-hal-uart.c shall not be used by the application.
This is a middleware to connect  IDF to Arduino API.

mrengineer7777 commented:
@SuGlider Apologies for not replying sooner.  Hectic day at work yesterday.  Your explanation on Github review buttons is very helpful.

Just read through all the code changes again.  With your latest commits PR looks good to me.

SuGlider commented:
> @SuGlider Apologies for not replying sooner. Hectic day at work yesterday. Your explanation on Github review buttons is very helpful.
> 
> Just read through all the code changes again. With your latest commits PR looks good to me.

@mrengineer7777 - Thank you for the code review!
